### ISSUE TITLE
Timeline of the next major release

### ISSUE THREAD

**Issue by wferi** [ClusterLabs/libqb#336]
Hi,

Are there any plans to release libqb 2 in the near future?
It has little chance of reaching Debian buster unless at least an RC is cut this week, but I have to ask because we experience strange problems with Pacemaker, see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=923246. Strangely enough, libqb itself and Corosync build fine, and binutils upstream failed to immediately reproduce the issue, but I expect a hard time arguing with the binutils maintainer, how blames libqb at the moment, because I'm not familiar with the "linker magic" libqb 1 employs (#314 may be related).

**Comment by wferi** [comment]
`libqb.pc` states:

> when concerned about a warning coming from the build process like
> warning: [...]libqb.so contains output sections; did you forget -T?
> and the build finishes OK, take it merely as a harmless side-effect

This used to be the case, but what is one supposed to do if the build fails? :)

**Comment by jnpkrn** [comment]
Will look into that if time permits, but what I see *a must* for the most
immediate new release -- long story short (still recovering from scars
on my scientifically-leaning soul) -- is to enforce _somehow_ that all the
arguments passed into `qb_logt` but the actual format interpolation ones
need to be (with ongoing contingency, see below) compile-time constants.

Items-split rationale is as follows:

1. for the state prior to the linker section force removal (incl. last
   officially released version of libqb), it's reasonable to expect that
   all the significant logging clients were compiled also for Linux,
   hence with linker section configure-time enabled in libqb and
   consequently via `qblog.h`

2. previous point enforced the very same compile-time constants
   restriction automatically (even though with a cryptic message
   from the compilation, IIRC), so the existing codebase can be
   considered strict in this regard

3. there are justified (since linker section feature was actually
   a very reasonably engineered run-time resource saver, not sure if
   it is Angus or Steven who deserves a pat on the back for that or
   whoever brought that idea initially) comments now that we shall
   do better to deal with the overhead that's pushed upon the
   logging clients now in the new world order:
   https://lists.clusterlabs.org/pipermail/developers/2019-January/001342.html
   (I myself have some niche optimizations in the sleeve, subject to
   testing if they actually help, but they are not relying on this
   prerequisite)

4. if new libqb version is to be cut tomorrow, we definitely want to
   keep an opportunity to optimize the logging in the future and some
   of that (the extent is very limited!) relies on retaining such
   a strict constraint, so having a release where this gets
   intermittently relaxed, only to be made strict again is
   a terrible disruption we shall rather avoid

Now, implementation-wise, it can be achieved with a little bit of
refactoring so that the actual `qb_log_callsite` struct is stored
statically, or there are various preprocessor/compiler tricks, see
the relatively recent addition to the Linux kernel (`__is_constexpr`):

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3c8ba0d61d04ced9f8d9ff93977995a9e4e96e91

If we have more time, there's indeed quite some wiggle room to piggy
back possibly breaking changes as the ABI contingency is doomed to
get broken, anyway.

**Comment by jnpkrn** [comment]
> or there are various preprocessor/compiler tricks

Note that this is also _a must_ should the Lars' patch be accepted, since
the system will behave treacherously otherwise, without a prior warning!

**Comment by wferi** [comment]
There's a good chance that binutils upstream [saved the day](https://sourceware.org/bugzilla/show_bug.cgi?id=24276). I couldn't test it, because even though the Debian binutils maintainer was *very* quick to cherry-pick and upload the fix, the necessary changes haven't trickled through the autobuild machinery yet. But the fact that binutils now has a libqb-inspired unit test is fairly reassuring. Maybe `gold` support also arrives one day...
In short, there's probably no need to rush the new release for this reason.

**Comment by feist** [comment]
@wferi unfortunately we don't have any definite plans for when we will be completing the 2.0 release, but as of right now it won't be for at *least* a few months, things may change in the future and we will keep you updated, but I would plan on using the current version of libqb for the Debian buster.

**Comment by wferi** [comment]
Thanks for the info, @feist! I've just tested the binutils fix, it enables building Pacemaker again, so we're good. I appreciate future updates, though there's no need to address me personally, asking for testing pre-releases on users@clusterlabs.org in probably more effective.
Thanks for the whole team for providing great free software!

**Comment by jnpkrn** [comment]
One of other ABI breaking ideas would be to make the poll handlers for
IPC server receive either self-reference to the service, or if more
encapsulation (less fragility towards the space of possible mishandlings)
is needed, at least the user pointer possibly previously associated with
`qb_ipcs_service_context_set` since doing so could possibly prevent from
having (undesirably) a global for the _mainloop_ around.

**Comment by chrissie-c** [comment]
Closing this as v2.0.0 was released on 4th May
